File: web/frontend/src/api/bedolaga.ts
import client from './client'

// ── Types ────────────────────────────────────────────────────────

export interface BedolagaStatus {
  enabled: boolean
  connected: boolean
  base_url: string | null
  bot_version: string | null
  sync_running: boolean
  initial_sync_done: boolean
  sync_entities: Array<{
    key: string
    last_sync_at: string | null
    sync_status: string
    records_synced: number
    error_message: string | null
  }>
}

export interface BedolagaOverview {
  total_users: number
  active_subscriptions: number
  total_revenue: number
  total_transactions: number
  open_tickets: number
  snapshot_at: string | null
  raw_data: Record<string, unknown> | null
}

export interface BedolagaUser {
  id: number
  telegram_id: number | null
  username: string | null
  first_name: string | null
  last_name: string | null
  status: string | null
  balance_rubles: number
  referral_code: string | null
  has_had_paid_subscription: boolean
  created_at: string | null
  last_activity: string | null
  synced_at: string | null
}

export interface BedolagaSubscription {
  id: number
  user_id: number | null
  user_telegram_id: number | null
  plan_name: string | null
  status: string | null
  is_trial: boolean
  started_at: string | null
  expires_at: string | null
  traffic_limit_bytes: number | null
  traffic_used_bytes: number | null
  payment_amount: number | null
  payment_provider: string | null
  synced_at: string | null
}

export interface BedolagaTransaction {
  id: number
  user_id: number | null
  user_telegram_id: number | null
  amount: number
  currency: string | null
  provider: string | null
  status: string | null
  type: string | null
  created_at: string | null
  synced_at: string | null
}

export interface BedolagaTransactionStats {
  total_amount: number
  total_count: number
  by_provider: Record<string, { amount: number; count: number }>
  by_day: Array<{ day: string; amount: number; count: number }>
}

export interface BedolagaRevenue {
  total_revenue: number
  revenue_today: number
  revenue_week: number
  revenue_month: number
  by_provider: Record<string, number>
  daily_chart: Array<{ day: string; amount: number }>
}

export interface PaginatedResponse<T> {
  items: T[]
  total: number
  limit: number
  offset: number
}

// ── API ──────────────────────────────────────────────────────────

export const bedolagaApi = {
  // Status & Health
  getStatus: async (): Promise<BedolagaStatus> => {
    const { data } = await client.get('/bedolaga/status')
    return data
  },

  // Sync
  triggerSync: async (entity?: string) => {
    const params = entity ? { entity } : {}
    const { data } = await client.post('/bedolaga/sync', null, { params })
    return data
  },

  getSyncStatus: async () => {
    const { data } = await client.get('/bedolaga/sync/status')
    return data
  },

  // Overview (from cache)
  getOverview: async (): Promise<BedolagaOverview> => {
    const { data } = await client.get('/bedolaga/overview')
    return data
  },

  getOverviewHistory: async (limit = 30) => {
    const { data } = await client.get('/bedolaga/overview/history', { params: { limit } })
    return data
  },

  // Users (from cache)
  getUsers: async (params: {
    limit?: number
    offset?: number
    status?: string
    search?: string
  } = {}): Promise<PaginatedResponse<BedolagaUser>> => {
    const { data } = await client.get('/bedolaga/users', { params })
    return {
      items: Array.isArray(data?.items) ? data.items : [],
      total: data?.total || 0,
      limit: data?.limit || 50,
      offset: data?.offset || 0,
    }
  },

  getUser: async (userId: number): Promise<BedolagaUser> => {
    const { data } = await client.get(`/bedolaga/users/${userId}`)
    return data
  },

  // Subscriptions (from cache)
  getSubscriptions: async (params: {
    limit?: number
    offset?: number
    status?: string
    user_id?: number
    is_trial?: boolean
  } = {}): Promise<PaginatedResponse<BedolagaSubscription>> => {
    const { data } = await client.get('/bedolaga/subscriptions', { params })
    return {
      items: Array.isArray(data?.items) ? data.items : [],
      total: data?.total || 0,
      limit: data?.limit || 50,
      offset: data?.offset || 0,
    }
  },

  // Transactions (from cache)
  getTransactions: async (params: {
    limit?: number
    offset?: number
    user_id?: number
    provider?: string
    status?: string
    type?: string
  } = {}): Promise<PaginatedResponse<BedolagaTransaction>> => {
    const { data } = await client.get('/bedolaga/transactions', { params })
    return {
      items: Array.isArray(data?.items) ? data.items : [],
      total: data?.total || 0,
      limit: data?.limit || 50,
      offset: data?.offset || 0,
    }
  },

  getTransactionStats: async (): Promise<BedolagaTransactionStats> => {
    const { data } = await client.get('/bedolaga/transactions/stats')
    return data
  },

  // Revenue (computed from cache)
  getRevenue: async (): Promise<BedolagaRevenue> => {
    const { data } = await client.get('/bedolaga/revenue')
    return data
  },

  // Tickets (real-time)
  getTickets: async (params: {
    limit?: number
    offset?: number
    status?: string
    priority?: string
    user_id?: number
  } = {}) => {
    const { data } = await client.get('/bedolaga/tickets', { params })
    return data
  },

  getTicket: async (ticketId: number) => {
    const { data } = await client.get(`/bedolaga/tickets/${ticketId}`)
    return data
  },

  updateTicketStatus: async (ticketId: number, status: string) => {
    const { data } = await client.post(`/bedolaga/tickets/${ticketId}/status`, { status })
    return data
  },

  replyToTicket: async (ticketId: number, text: string) => {
    const { data } = await client.post(`/bedolaga/tickets/${ticketId}/reply`, { text })
    return data
  },

  // Promo Codes (real-time)
  getPromoCodes: async (params: { limit?: number; offset?: number; is_active?: boolean } = {}) => {
    const { data } = await client.get('/bedolaga/promo-codes', { params })
    return data
  },

  createPromoCode: async (payload: Record<string, unknown>) => {
    const { data } = await client.post('/bedolaga/promo-codes', payload)
    return data
  },

  deletePromoCode: async (promocodeId: number) => {
    await client.delete(`/bedolaga/promo-codes/${promocodeId}`)
  },

  // Partners (real-time)
  getPartners: async (params: { limit?: number; offset?: number; search?: string } = {}) => {
    const { data } = await client.get('/bedolaga/partners', { params })
    return data
  },

  getPartnerStats: async (days = 30) => {
    const { data } = await client.get('/bedolaga/partners/stats', { params: { days } })
    return data
  },

  // Polls (real-time)
  getPolls: async (params: { limit?: number; offset?: number } = {}) => {
    const { data } = await client.get('/bedolaga/polls', { params })
    return data
  },

  // Broadcasts (real-time)
  getBroadcasts: async (params: { limit?: number; offset?: number } = {}) => {
    const { data } = await client.get('/bedolaga/broadcasts', { params })
    return data
  },

  // Settings (real-time)
  getSettings: async () => {
    const { data } = await client.get('/bedolaga/settings')
    return data
  },

  updateSetting: async (key: string, value: unknown) => {
    const { data } = await client.put(`/bedolaga/settings/${key}`, { value })
    return data
  },
}

File: web/backend/schemas/bedolaga.py
"""Pydantic schemas for Bedolaga integration endpoints."""

from datetime import datetime
from typing import Any, Optional, List

from pydantic import BaseModel, Field


# ── Sync & Health ─────────────────────────────────────────────────

class BedolagaStatusResponse(BaseModel):
    enabled: bool
    connected: bool
    base_url: Optional[str] = None
    bot_version: Optional[str] = None
    last_sync_at: Optional[datetime] = None
    sync_status: Optional[str] = None


class BedolagaSyncStatusResponse(BaseModel):
    entity: str
    last_sync_at: Optional[datetime] = None
    status: str = "never"
    records_synced: int = 0
    error_message: Optional[str] = None


class BedolagaSyncResult(BaseModel):
    entities: List[BedolagaSyncStatusResponse]
    triggered_at: datetime


# ── Stats Snapshot ────────────────────────────────────────────────

class BedolagaOverviewResponse(BaseModel):
    total_users: int = 0
    active_subscriptions: int = 0
    total_revenue: float = 0.0
    total_transactions: int = 0
    open_tickets: int = 0
    snapshot_at: Optional[datetime] = None
    raw_data: Optional[dict] = None


# ── Users (synced) ────────────────────────────────────────────────

class BedolagaUserResponse(BaseModel):
    id: int
    telegram_id: Optional[int] = None
    username: Optional[str] = None
    first_name: Optional[str] = None
    last_name: Optional[str] = None
    status: Optional[str] = None
    balance_rubles: float = 0.0
    referral_code: Optional[str] = None
    has_had_paid_subscription: bool = False
    created_at: Optional[datetime] = None
    last_activity: Optional[datetime] = None
    synced_at: Optional[datetime] = None


class BedolagaUserListResponse(BaseModel):
    items: List[BedolagaUserResponse]
    total: int
    limit: int
    offset: int


# ── Subscriptions (synced) ────────────────────────────────────────

class BedolagaSubscriptionResponse(BaseModel):
    id: int
    user_id: Optional[int] = None
    user_telegram_id: Optional[int] = None
    plan_name: Optional[str] = None
    status: Optional[str] = None
    is_trial: bool = False
    started_at: Optional[datetime] = None
    expires_at: Optional[datetime] = None
    traffic_limit_bytes: Optional[int] = None
    traffic_used_bytes: Optional[int] = None
    payment_amount: Optional[float] = None
    payment_provider: Optional[str] = None
    synced_at: Optional[datetime] = None


class BedolagaSubscriptionListResponse(BaseModel):
    items: List[BedolagaSubscriptionResponse]
    total: int
    limit: int
    offset: int


# ── Transactions (synced) ─────────────────────────────────────────

class BedolagaTransactionResponse(BaseModel):
    id: int
    user_id: Optional[int] = None
    user_telegram_id: Optional[int] = None
    amount: float = 0.0
    currency: Optional[str] = None
    provider: Optional[str] = None
    status: Optional[str] = None
    type: Optional[str] = None
    created_at: Optional[datetime] = None
    synced_at: Optional[datetime] = None


class BedolagaTransactionListResponse(BaseModel):
    items: List[BedolagaTransactionResponse]
    total: int
    limit: int
    offset: int


class BedolagaTransactionStatsResponse(BaseModel):
    total_amount: float = 0.0
    total_count: int = 0
    by_provider: Optional[dict] = None
    by_day: Optional[List[dict]] = None
    period_from: Optional[datetime] = None
    period_to: Optional[datetime] = None


# ── Tickets (real-time) ──────────────────────────────────────────

class BedolagaTicketResponse(BaseModel):
    id: int
    user_id: Optional[int] = None
    user_telegram_id: Optional[int] = None
    username: Optional[str] = None
    status: Optional[str] = None
    priority: Optional[str] = None
    subject: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    messages: Optional[List[dict]] = None


class BedolagaTicketListResponse(BaseModel):
    items: List[BedolagaTicketResponse]
    total: int
    limit: int
    offset: int


class BedolagaTicketStatusUpdate(BaseModel):
    status: str = Field(..., description="New ticket status")


class BedolagaTicketPriorityUpdate(BaseModel):
    priority: str = Field(..., description="New ticket priority")


class BedolagaTicketReply(BaseModel):
    text: str = Field(..., min_length=1, max_length=4000)


# ── Promo Groups (real-time) ─────────────────────────────────────

class BedolagaPromoGroupResponse(BaseModel):
    id: int
    name: Optional[str] = None
    description: Optional[str] = None
    member_count: int = 0
    created_at: Optional[datetime] = None
    raw_data: Optional[dict] = None


class BedolagaPromoGroupListResponse(BaseModel):
    items: List[BedolagaPromoGroupResponse]
    total: int
    limit: int
    offset: int


class BedolagaPromoGroupCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = None


class BedolagaPromoGroupUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = None


# ── Promo Codes (real-time) ──────────────────────────────────────

class BedolagaPromoCodeResponse(BaseModel):
    id: int
    code: Optional[str] = None
    is_active: bool = True
    discount_percent: Optional[float] = None
    discount_amount: Optional[float] = None
    max_uses: Optional[int] = None
    used_count: int = 0
    expires_at: Optional[datetime] = None
    created_at: Optional[datetime] = None
    raw_data: Optional[dict] = None


class BedolagaPromoCodeListResponse(BaseModel):
    items: List[BedolagaPromoCodeResponse]
    total: int
    limit: int
    offset: int


class BedolagaPromoCodeCreate(BaseModel):
    code: str = Field(..., min_length=1, max_length=50)
    discount_percent: Optional[float] = Field(None, ge=0, le=100)
    discount_amount: Optional[float] = Field(None, ge=0)
    max_uses: Optional[int] = Field(None, ge=1)
    expires_at: Optional[datetime] = None


class BedolagaPromoCodeUpdate(BaseModel):
    is_active: Optional[bool] = None
    discount_percent: Optional[float] = Field(None, ge=0, le=100)
    discount_amount: Optional[float] = Field(None, ge=0)
    max_uses: Optional[int] = Field(None, ge=1)
    expires_at: Optional[datetime] = None


# ── Polls (real-time) ────────────────────────────────────────────

class BedolagaPollResponse(BaseModel):
    id: int
    title: Optional[str] = None
    description: Optional[str] = None
    is_active: bool = False
    questions: Optional[List[dict]] = None
    created_at: Optional[datetime] = None
    raw_data: Optional[dict] = None


class BedolagaPollListResponse(BaseModel):
    items: List[BedolagaPollResponse]
    total: int
    limit: int
    offset: int


# ── Partners (real-time) ─────────────────────────────────────────

class BedolagaPartnerResponse(BaseModel):
    user_id: int
    username: Optional[str] = None
    telegram_id: Optional[int] = None
    referral_count: int = 0
    total_earned: float = 0.0
    commission_rate: Optional[float] = None
    raw_data: Optional[dict] = None


class BedolagaPartnerListResponse(BaseModel):
    items: List[BedolagaPartnerResponse]
    total: int
    limit: int
    offset: int


class BedolagaPartnerStatsResponse(BaseModel):
    total_referrers: int = 0
    total_referrals: int = 0
    total_earned: float = 0.0
    daily_stats: Optional[List[dict]] = None
    raw_data: Optional[dict] = None


# ── Revenue Analytics (computed from synced data) ─────────────────

class BedolagaRevenueResponse(BaseModel):
    total_revenue: float = 0.0
    revenue_today: float = 0.0
    revenue_week: float = 0.0
    revenue_month: float = 0.0
    by_provider: Optional[dict] = None
    daily_chart: Optional[List[dict]] = None


# ── Generic passthrough ──────────────────────────────────────────

class BedolagaProxyResponse(BaseModel):
    """Generic response for passthrough proxy endpoints."""
    data: Any = None
    status: str = "ok"
